skymap_scanner/server/start_scan.py

```diff
@@ -419,30 +499,78 @@ async def collect(
             if best.nside not in self.nsides_dict:
                 self.nsides_dict[best.nside] = {}
             if best.pixel in self.nsides_dict[best.nside]:
-                raise DuplicatePixelRecoException(
+                raise ExtraPixelRecoException(
                     f"NSide {best.nside} / Pixel {best.pixel} is already in nsides_dict"
                 )
             self.nsides_dict[best.nside][best.pixel] = best
             LOGGER.debug(f"Saved (found during {logging_id}): {best.id_tuple} {best}")
 
         # report after potential save
-        await self.reporter.record_pixreco(pixreco.nside, pixreco_start, pixreco_end)
+        await self.reporter.record_pixreco(
+            pixreco.nside,
+            pixreco_runtime,
+            roundtrip_start=sent_pixel.sent_time,
+            roundtrip_end=time.time(),
+        )
 
+    def is_scan_done(self) -> bool:
+        """Has every pixel been collected?"""
+        # first check lengths, faster: O(1)
+        if self.n_sent != len(self._pixrecoid_received_quick_lookup):
+            return False
+        # now, sanity check contents, slower: O(n)
+        sent_ids = set((p.nside, p.pixel_id, p.posvar_id) for p in self.sent_pixels)
+        if sent_ids == self._pixrecoid_received_quick_lookup:
+            return True
+        raise RuntimeError(
+            f"Sanity check failed: Collected enough pixels,"
+            f" but does not match: {sent_ids=} vs {self._pixrecoid_received_quick_lookup=}"
+        )
+
+    def ok_to_serve_more(self) -> bool:
+        """Return whether enough pixel-recos collected to serve more.
 
-async def serve(
+        If we are approaching the end of the scan, always return False.
+        """
+        if self._end_game:
+            return False
+
+        if not self.nsides_dict:  # nothing has been saved yet
+            return False
+
+        # get percentage of latest nside (otherwise BIG + small >> threshold)
+        latest_saved_nside = max(self.nsides_dict.keys())
+
+        # Have we already said yes to this nside?
+        # Ex: nside=8 reached threshold so we sent out nside=64,
+        #     now we're asking about nside=8 again--it's not okay to serve more.
+        if latest_saved_nside < max(self._sent_pixels_by_nside.keys()):
+            return False
```

Comment:
I'm not so sure this should be `False`. Since we start the next iteration before the current one is complete, it may be that there is a better reco found after the threshold was first crossed. So I think we'd want to run the refinement code each time a new pixel is received, even if it is for a lower nside. Does that make sense?